### PR TITLE
Introduce ObjLogger

### PR DESCRIPTION
**Pull request by patuwwy**
This PR introduces Object Logger (ObjLogger) and replaces old plain text Logger. 

The main assumption of the object logger is to send logs in the form of JSON objects. ObjLogger also enables the transfer of logs from one logger to another and automatic enrichment of logs with information about the components through which the logs pass.

image

Added option to filter logs showing in terminal

```
Usage: sth [options]

Options:
  -L, --log-level <level>         Specify log level (default: "trace")
```

### CODE REVIEW

File: packages/host/src/lib/host.ts [hunk 1]
```diff
@@ -72,6 +76,11 @@ export class Host implements IComponent {
      */
     logger: Logger;
 
+    /**
+     * Instance of class providing logging utilities.
+     */
+    objLogger: ObjLogger;
```
Comment (f1ames) [inline packages/host/src/lib/host.ts:82]: Would `logger` be still used for something, e.g. `/log` endpoint or we plan to get rid of it entirely?

**Review by f1ames — CHANGES_REQUESTED** [review]
Logs look really crisp now! :star_struck: Good job :+1: 

image

As for clarifications:

For me the question is if we want to keep two separate loggers or merge them into one and then control per output stream how something is logged (json vs text). Or just get rid of `Logger` since `objLogger` already outputs text to stdout :thinking:  Maybe it's not for now :thinking: but still worth considering.
---

There are some classes where (`process-sequence-adapter.ts` and `sequence-adapter.ts`):

```
objLogger: IObjectLogger;
```

is used and others where it's like:

```
objLogger: ObjLogger;
```

Minor thing but seems inconsistent :thinking:

File: packages/obj-logger/src/obj-logger.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+                this.write(entry.level!, entry, ...entry.data || []);
+            });
+
+        this.outLogStream.pipe(this.output);
+    }
+
+    write(level: LogLevel, entry: LogEntry | string, ...optionalParams: any[]) {
+        if (ObjLogger.levels.indexOf(level) <= ObjLogger.levels.indexOf(this.logLevel)) {
+            //return;
```
[the remark this answers is not in the document]
Comment (patuwwy) [reply packages/obj-logger/src/obj-logger.ts:51]: yup :)

**Comment by patuwwy** [comment]
@f1ames, ty for cr
previous logger is going to be totally removed / replaced with the new one.
this PR is still missing a commit removing all `this.logger.log`

File: packages/obj-logger/src/obj-logger.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    error(entry: LogEntry | string, ...optionalParams: any[]) {
+        this.write("ERROR", entry, ...optionalParams);
+    }
+
+    debug(entry: LogEntry | string, ...optionalParams: any[]) {
+        this.write("DEBUG", entry, ...optionalParams);
+    }
+
+    warn(entry: LogEntry | string, ...optionalParams: any[]) {
```
[the remark this answers is not in the document]
Comment (patuwwy) [reply packages/obj-logger/src/obj-logger.ts:101]: fatal is actually missing, thx

File: packages/types/src/object-logger.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    data?: any[];
+    from: string;
+}>
+
+export interface IObjectLogger {
+    write(level: LogEntry["level"], entry: LogEntry | string, ...optionalParams: any[]): void;
+    trace(entry: LogEntry | string, ...optionalParams: any[]): void;
+    debug(entry: LogEntry | string, ...optionalParams: any[]): void;
+    info(entry: LogEntry | string): void
+    pipe(target: Writable | IObjectLogger): void;
```
[the remark this answers is not in the document]
Comment (patuwwy) [reply packages/types/src/object-logger.ts:21]: yes , and fatal as well

File: packages/obj-logger/src/obj-logger.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
+/* eslint-disable no-console */
+import { DataStream, StringStream } from "scramjet";
+import { PassThrough, Writable } from "stream";
+import { IObjectLogger, LogEntry, LogLevel } from "@scramjet/types";
+import { getName } from "./utils/get-name";
+
+export class ObjLogger implements IObjectLogger {
+    outLogStream = new PassThrough({ objectMode: true });
+    inLogStream = new PassThrough({ objectMode: true });
+    inStringStream = new PassThrough({ objectMode: false });
```
Comment (alicja-gruzdz) [inline packages/obj-logger/src/obj-logger.ts:10]: I think we chatted about it but it escaped me, can you please remind why it was introduced?

**Review by alicja-gruzdz — APPROVED** [review]

File: packages/host/src/lib/host.ts (hunk 1 above)
Comment (patuwwy) [reply packages/host/src/lib/host.ts:82]: get rid. `/log` now provides logs in nd-json format

File: packages/obj-logger/src/obj-logger.ts (hunk 5 above)
Comment (patuwwy) [reply packages/obj-logger/src/obj-logger.ts:10]: i will add comments to this class :)

**Comment by patuwwy** [comment]
> ```
> objLogger: IObjectLogger;
> ```
> 
> is used and others where it's like:
> 
> ```
> objLogger: ObjLogger;
> ```
> 
> Minor thing but seems inconsistent 🤔

done

File: packages/adapters/src/dockerode-docker-helper.ts [hunk 6]
```diff
@@ -184,14 +184,14 @@ export class DockerodeDockerHelper implements IDockerHelper {
...
-            this.logger.log("Start pulling image", name);
+            this.objLogger.trace("Start pulling image", name);
 
             const pullStream = await this.dockerode.pull(name);
 
             // Wait for pull to finish
             await new Promise(res => this.dockerode.modem.followProgress(pullStream, res));
 
-            this.logger.log("Pulling image", name, "done.");
+            this.objLogger.trace("Image pulled");
```
Comment (MichalCz) [inline packages/adapters/src/dockerode-docker-helper.ts:194]:
```suggestion
            this.objLogger.trace("Image pulled", name);
```

File: packages/obj-logger/src/obj-logger.ts [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    constructor(reference: any, baseLog: LogEntry = {}, logLevel: LogLevel = "TRACE") {
+        this.name = getName(reference);
+
+        this.baseLog = baseLog;
+        this.logLevel = logLevel;
+
+        StringStream
+            .from(this.inputStringifiedLogStream)
+            .JSONParse(true)
+            .pipe(this.inputLogStream);
```
Comment (MichalCz) [inline packages/obj-logger/src/obj-logger.ts:67]: Do we need to parse this every time? We could just mux our log into the others. It's ok to leave it like this, but maybe a tech debt entry would be in order?

File: packages/obj-logger/src/utils/pretty-print.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+    DEBUG: COLORS.FgBlue,
+    INFO: COLORS.FgCyan,
+    WARN: COLORS.FgYellow,
+    ERROR: COLORS.FgRed,
+    FATAL: COLORS.FgMagenta
+};
+
+const getLevelColor = (level: LogLevel = "INFO") => `${COLOR_MAP[level]}${level?.padEnd(5)}${COLORS.Reset}`;
+
+export const prettyPrint = (obj: LogEntry) => {
```
Comment (MichalCz) [inline packages/obj-logger/src/utils/pretty-print.ts:15]: Couldn't find any references in the code? Shouldn't this be in cli?

Comment (MichalCz) [inline packages/obj-logger/src/utils/pretty-print.ts:15]: We should have another version with no color if color was not turned on, or logs are sent to non tty.

Comment (MichalCz) [reply packages/obj-logger/src/utils/pretty-print.ts:15]: https://github.com/chalk/supports-color

Some inspiration...

File: packages/types/src/sth-configuration.ts [hunk 9]
```diff
@@ -68,6 +70,11 @@ export type HostConfig = {
 }
 
 export type STHConfiguration = {
+    /**
+     * Logging level.
+     */
+    logLevel: LogLevel
```
Comment (MichalCz) [inline packages/types/src/sth-configuration.ts:76]: `logColor` maybe

File: packages/obj-logger/src/obj-logger.ts (hunk 7 above)
Comment (patuwwy) [reply packages/obj-logger/src/obj-logger.ts:67]: In the present solution, we parse incoming logs to retrieve some of the data from them and complete the data of the logger they come to.

inputStringifiedLogStream is used only when stringified data are sent, for example via http (runner -> instance-adapter)

File: packages/types/src/sth-configuration.ts (hunk 9 above)
Comment (patuwwy) [reply packages/types/src/sth-configuration.ts:76]: added

Comment (f1ames) [reply packages/types/src/sth-configuration.ts:76]: :thinking:

File: packages/obj-logger/src/utils/pretty-print.ts (hunk 8 above)
Comment (patuwwy) [reply packages/obj-logger/src/utils/pretty-print.ts]: this is used in host https://github.com/scramjetorg/transform-hub/blob/feature/obj-logger/packages/host/src/lib/host.ts#L120

**Review by f1ames — APPROVED** [review]
Looking good :+1: :rainbow: 

image

---

Since `objLogger` is now the only logger (in most places) maybe we can call all it's instances `logger` (I see it's already done in some places), I mean in classes like `this.logger`? A minor thing and it would require renaming in all places so consider it a suggestion only :smile:

File: packages/obj-logger/src/obj-logger.ts (hunk 7 above)
Comment (MichalCz) [reply packages/obj-logger/src/obj-logger.ts:67]: Ok, my comment is not applicable in that case.

**Comment by patuwwy** [comment]
@f1ames all `objLogger` fields renamed to `logger`

**Pull request merged**